File: app/src/main/java/com/marked/vifo/activity/SignUpActivity.java
package com.marked.vifo.activity;

import android.app.ProgressDialog;
import android.content.IntentFilter;
import android.graphics.Color;
import android.os.Bundle;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.Snackbar;
import android.support.v4.app.FragmentManager;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.widget.Toast;

import com.android.volley.AuthFailureError;
import com.android.volley.NetworkError;
import com.android.volley.NetworkResponse;
import com.android.volley.NoConnectionError;
import com.android.volley.ParseError;
import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.ServerError;
import com.android.volley.TimeoutError;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.JsonRequest;
import com.marked.vifo.R;
import com.marked.vifo.controller.RequestQueue;
import com.marked.vifo.extra.HTTPStatusCodes;
import com.marked.vifo.fragment.SignUpEmailFragment;
import com.marked.vifo.fragment.SignUpNameFragment;
import com.marked.vifo.fragment.SignUpPassFragment;
import com.marked.vifo.gcm.RegistrationBroadcastReceiver;
import com.marked.vifo.extra.GCMConstants;
import com.marked.vifo.gcm.service.LogInRegistrationIntentService;
import com.marked.vifo.helper.Utils;

import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;

public class SignUpActivity extends AppCompatActivity implements SignUpNameFragment.SignUpNameFragmentInteraction, SignUpEmailFragment.SignUpEmailFragmentInteraction, SignUpPassFragment.SignUpPassFragmentInteraction {
    private FragmentManager mFragmentManager;
    private RegistrationBroadcastReceiver mRegistrationBroadcastReceiver;
    private CoordinatorLayout mContainer;
    private LocalBroadcastManager mBroadcastManagerastManager;
    private String mName;
    private String mEmail;
    private String mPassword;
    private RequestQueue mRequestQueue;
    private ProgressDialog mProgressDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sign_up);
        mContainer = (CoordinatorLayout) findViewById(R.id.fragmentContainer);
        mProgressDialog = new ProgressDialog(this);

        mFragmentManager = getSupportFragmentManager();
        mBroadcastManagerastManager = LocalBroadcastManager.getInstance(this);
        mRequestQueue = RequestQueue.getInstance(this);
        // GCM registration
        mRegistrationBroadcastReceiver = new RegistrationBroadcastReceiver();
        mRegistrationBroadcastReceiver.setOnRegistrationBroadcastReceiverListener(new RegistrationBroadcastReceiver.RegistrationBroadcastReceiverListener() {
            @Override
            public void onDismiss() {
                mProgressDialog.dismiss();
            }

            @Override
            public void onError(int errorStatusCode) {
                mProgressDialog.dismiss();
                Snackbar snackbar = null;
                switch (errorStatusCode) {
                    case HTTPStatusCodes.REQUEST_CONFLICT:
                        snackbar = Utils.createWhiteSnackBar(SignUpActivity.this, mContainer, "You already have an account");
                        break;
                    case HTTPStatusCodes.REQUEST_TIMEOUT:
                        snackbar = Utils.createWhiteSnackBar(SignUpActivity.this, mContainer, "Timeout error");
                        break;
                    case HTTPStatusCodes.REQUEST_INCORRECT_PASSWORD:
                        snackbar = Utils.createWhiteSnackBar(SignUpActivity.this, mContainer, "Incorrect password");
                        break;
                    case HTTPStatusCodes.NOT_FOUND:
                        snackbar = Utils.createWhiteSnackBar(SignUpActivity.this, mContainer, "We are sorry, but we did not found you");
                        break;
                }
                if (snackbar != null) {
                    snackbar.getView().setBackgroundColor(Color.WHITE);
                    snackbar.show();
                }
            }
        });

        mFragmentManager.beginTransaction().add(R.id.fragmentContainer, SignUpNameFragment.newInstance()).commit();
    }

    @Override
    public void onResume() {
        super.onResume();
        mBroadcastManagerastManager.registerReceiver(mRegistrationBroadcastReceiver, new IntentFilter(GCMConstants.ACTION_SIGNUP));
        mBroadcastManagerastManager.registerReceiver(mRegistrationBroadcastReceiver, new IntentFilter(GCMConstants.ACTION_ERROR));
    }

    @Override
    protected void onPause() {
        LocalBroadcastManager.getInstance(this).unregisterReceiver(mRegistrationBroadcastReceiver);
        super.onPause();
    }

    @Override
    public void onSaveName(String name) {
        mName = name;
        mFragmentManager.beginTransaction().add(R.id.fragmentContainer, SignUpEmailFragment.newInstance()).addToBackStack(null).commit();
    }

    @Override
    public void onSaveEmail(String email) {
        mProgressDialog.show();
        mEmail = email;
        checkEmail(mEmail);
    }

    @Override
    public void onPasswordSave(String password) {
        mPassword = password;
        LogInRegistrationIntentService.startActionSignup(this,mName, mEmail, mPassword);
        mProgressDialog = ProgressDialog.show(this, "Signup", "Please wait ...", true);
    }

    /**
     * Sends a request to the server to check if the email already exists.
     * If the server has the email, the user already has an account and we should tell him that
     * Else proceed to the next step
     *
     * @param email the email adress to send to the server
     */
    private void checkEmail(String email) {
        String verifyUserURL = null;
        try {
            verifyUserURL = GCMConstants.SERVER_USER_EXISTS +
                            "?email=" + URLEncoder.encode(email, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        JsonRequest jsonRequest = new JsonObjectRequest(Request.Method.GET, verifyUserURL, new Response.Listener<JSONObject>() {
            @Override
            public void onResponse(JSONObject response) {
                mProgressDialog.dismiss();
                mFragmentManager.beginTransaction().add(R.id.fragmentContainer, SignUpPassFragment.newInstance()).addToBackStack(null).commit();
            }
        }, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
                mProgressDialog.dismiss();
                mEmail = null;
                NetworkResponse networkResponse = error.networkResponse;
                if (networkResponse != null) {
                    switch (networkResponse.statusCode) {
                        case HTTPStatusCodes.REQUEST_TIMEOUT:
                            Toast.makeText(SignUpActivity.this, "Timeout", Toast.LENGTH_LONG).show();
                            break;
                        case HTTPStatusCodes.REQUEST_CONFLICT:
                            Toast.makeText(SignUpActivity.this, "You already have an account", Toast.LENGTH_LONG).show();
                            break;
                        default:
                            break;
                    }
                    Log.e("Volley", "Error. HTTP Status Code:" + networkResponse.statusCode);
                }
                if (error instanceof TimeoutError) {
                    Log.e("Volley", "TimeoutError");
                } else if (error instanceof NoConnectionError) {
                    Log.e("Volley", "NoConnectionError");
                } else if (error instanceof AuthFailureError) {
                    Log.e("Volley", "AuthFailureError");
                } else if (error instanceof ServerError) {
                    Log.e("Volley", "ServerError");
                } else if (error instanceof NetworkError) {
                    Log.e("Volley", "NetworkError");
                } else if (error instanceof ParseError) {
                    Log.e("Volley", "ParseError");
                }
            }
        });

        mRequestQueue.add(jsonRequest);
    }


}

File: app/src/main/java/com/marked/vifo/activity/SignUpActivity.kt
package com.marked.vifo.activity

import android.app.ProgressDialog
import android.content.IntentFilter
import android.os.Bundle
import android.support.v4.content.LocalBroadcastManager
import android.support.v7.app.AppCompatActivity
import android.util.Log
import com.android.volley.*
import com.android.volley.toolbox.JsonObjectRequest
import com.marked.vifo.R
import com.marked.vifo.delegate.DialogRegistration
import com.marked.vifo.extra.GCMConstants
import com.marked.vifo.extra.HTTPStatusCodes
import com.marked.vifo.extra.ServerConstants
import com.marked.vifo.fragment.signup.SignUpEmailFragment
import com.marked.vifo.fragment.signup.SignUpNameFragment
import com.marked.vifo.fragment.signup.SignUpPassFragment
import com.marked.vifo.gcm.RegistrationBroadcastReceiver
import com.marked.vifo.gcm.service.LogInRegistrationIntentService
import com.marked.vifo.helper.Toast
import com.marked.vifo.helper.add
import com.marked.vifo.helper.addToBackStack
import com.marked.vifo.model.RequestQueue
import java.net.URLEncoder

class SignUpActivity : AppCompatActivity(), SignUpNameFragment.SignUpNameFragmentInteraction, SignUpEmailFragment.SignUpEmailFragmentInteraction, SignUpPassFragment.SignUpPassFragmentInteraction {
	private val mFragmentManager by lazy { supportFragmentManager }
	private val mProgressDialog by lazy { ProgressDialog(this) }
	private val mRegistrationBroadcastReceiver by lazy { RegistrationBroadcastReceiver(DialogRegistration(this, mProgressDialog)) }
	private val mBroadcastManagerastManager by lazy { LocalBroadcastManager.getInstance(this) }
	private val mRequestQueue by lazy { RequestQueue.getInstance(this) }

	private var mName: String? = null
	private var mEmail: String? = null
	private var mPassword: String? = null

	override
	fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_sign_up)

		mFragmentManager.add(R.id.fragmentContainer, SignUpNameFragment.newInstance())
	}

	override fun onResume() {
		super.onResume()
		mBroadcastManagerastManager.registerReceiver(mRegistrationBroadcastReceiver, IntentFilter(GCMConstants.ACTION_SIGNUP))
		mBroadcastManagerastManager.registerReceiver(mRegistrationBroadcastReceiver, IntentFilter(GCMConstants.ACTION_ERROR))
	}

	override
	fun onPause() {
		mBroadcastManagerastManager.unregisterReceiver(mRegistrationBroadcastReceiver)
		super.onPause()
	}

	override
	fun onSaveName(name: String) {
		mName = name
		mFragmentManager.addToBackStack(R.id.fragmentContainer, SignUpEmailFragment.newInstance(mName))
	}

	override fun onSaveEmail(email: String) {
		mProgressDialog.show()
		mEmail = email
		checkEmail(mEmail)
	}

	override fun onPasswordSave(password: String) {
		mPassword = password
		LogInRegistrationIntentService.startActionSignup(this, mName, mEmail, mPassword)
		mProgressDialog.setTitle("Signup")
		mProgressDialog.setMessage("Please wait...")
		mProgressDialog.isIndeterminate = true
		mProgressDialog.show()
	}

	/**
	 * Sends a request to the server to check if the email already exists.
	 * If the server has the email, the user already has an account and we should tell him that
	 * Else proceed to the next step

	 * @param email the email adress to send to the server
	 */
	private fun checkEmail(email: String?) {
		var verifyUserURL = "${ServerConstants.SERVER_USER_EXISTS}?email=${URLEncoder.encode(email, "UTF-8")}"

		val jsonRequest = JsonObjectRequest(Request.Method.GET, verifyUserURL, Response.Listener<org.json.JSONObject> {
			mProgressDialog.dismiss()
			mFragmentManager.addToBackStack(R.id.fragmentContainer, SignUpPassFragment.newInstance())
		}, Response.ErrorListener { error ->
			mProgressDialog.dismiss()
			mEmail = null
			val networkResponse = error.networkResponse
			when (networkResponse?.statusCode) {
				HTTPStatusCodes.REQUEST_TIMEOUT -> Toast("Timeout")
				HTTPStatusCodes.REQUEST_CONFLICT -> Toast("You already have an account")
			}
			Log.e("Volley", "Error. HTTP Status Code:" + networkResponse.statusCode)
			when (error) {
				is TimeoutError ->
					Log.e("Volley", "TimeoutError")
				is NoConnectionError ->
					Log.e("Volley", "NoConnectionError")
				is AuthFailureError ->
					Log.e("Volley", "AuthFailureError")
				is ServerError ->
					Log.e("Volley", "ServerError")
				is NetworkError ->
					Log.e("Volley", "NetworkError")
				is ParseError -> Log.e("Volley", "ParseError")
			}
		})

		mRequestQueue.add(jsonRequest)
	}


}

